src/asl.ts

```diff
@@ -3345,6 +3362,51 @@ export class ASL {
     );
   }
 
+  /**
+   * In some cases, variable can be batch declared as state parameters.
+   * This method is an alternative to {@link evalDecl} which can evaluate multiple {@link ParameterDecl}s
+   * with initial values into a parameter object and any state to be run after (which setup any bound variables.)
+   *
+   * @returns a tuple tuple
+   *          1. a {@link Parameters} object intended to be used in the Parameters or ResultSelector of a state.
+   *             Contains the initialized parameter names. Could be empty.
+   *          2. a state or sub-state used to generate the bound names. May be undefined.
+   */
+  public evalParameterDeclForStateParameter(
+    node: FunctionlessNode,
+    ...parameters: [ParameterDecl | undefined, ASLGraph.JsonPath][]
+  ): [
+    Record<string, Parameters>,
+    ASLGraph.NodeState | ASLGraph.SubState | undefined
+  ] {
+    const [params, states] = parameters.reduce(
+      (
+        [params, states]: [
+          Record<string, Parameters>,
+          (ASLGraph.SubState | ASLGraph.NodeState | undefined)[]
+        ],
+        [decl, jsonPath]
+      ) => {
+        if (!decl) {
+          return [params, states];
+        }
+
+        return isIdentifier(decl.name)
+          ? [
+              {
+                ...params,
+                [`${this.getIdentifierName(decl.name)}.$`]: jsonPath.jsonPath,
+              },
+              states,
+            ]
+          : [params, [...states, this.evalAssignment(decl.name, jsonPath)]];
+      },
+      [{}, []]
+    );
+
+    return [params, ASLGraph.joinSubStates(node, ...states)];
+  }
```

Comment:
Consider chunking this function up into a `map`/`flatMap` prior to the `reduce` - I find it hard to tease apart spreads from the evaluation logic.

Reply:
I think I can add another short pass to partition by `isIdentifier` now to clean this up.